Refactored login hint process

begedin opened the pull request:
Hopefully resolves #130 

Regarding the `createFromState` and `openFromState` methods, they are already implemented and seem to be working fine. The thing that did seem to change are the parameters we need to send through the URL. I updated the design_notes.txt (replaced wth an .md) with an explanation how to form the URL exactly.

Regarding extraction of `cacheLoginHint` into a library, I went a step further and also extracted the methods required for login hint retrieval (used in the authenticator) into the same library. This way, we keep that logic in one place.

begedin committed:
Refactored login hint process

begedin committed:
Fixed incorrect import

begedin committed:
Minor fixes

venkatd commented:
A few points:

-Currently the `load` method will cache the login hint but `openFromState` or `createFromState`--which are the methods used from the gdrive browser--won't.

-The first time someone double clicks on an file from the drive browser, it'll call `authorize`. Because the auth doesn't know which user is logging in, you'll get a popup asking them to pick an account. However, from the `state` parameter, we know the userId so we can pass in a `login_hint` to avoid this.

As a reminder, track these hours separately :)

begedin committed:
Design notes correction

begedin committed:
Moved login hint caching to afterModel hook

begedin committed:
Moved login hint fetching to auth library

begedin commented on addon/mixins/authenticated-route-mixin.js:
```diff
@@ -1,21 +1,15 @@
-        return doc;
-      });
+  afterModel: function (document, transition) {
+    var userId = this.get('session.secure.id');
```
I missed a change in `ember-simple-auth`. The payload returned by a successful authentication request is no longer stored directly in the session object. It's now stored as a `session.secure` property, so that's where the user id is now.

begedin commented:
@venkatd: I wasted some time trying to move it all into the document-source library, but eventually, I figured out a simple solution for both of the issues you described.

By moving the login hint caching from the `model` hook into `afterMode`, it is now called for any case, regardless of how the document was opened or created. While it was being called from model, it would only work with cases where there's a document id in the parameter list, meaning only direct load access through regular routing, since that's the only time the model hook get's executed.

For `open/createFromState`, we fetch or create the document first and then pass it as a model directly to the route, so the `model` hook is skipped. `afterModel` is never skipped, though, so that's where the logic really needed to be in the first place.

As for authorization, I simply moved the fetching of the login hint (and the `GOOGLE_CLIENT_ID` config property) directly into the `auth` library, since it always needs to be called so it makes more sense for it to be there than in the authenticator.

Since opening from google drive is functionally equal to opening from state, this change applies to that as well and there's an attempt to determine the login hint every time.

venkatd commented:
@begedin we would still need to handle figuring out the login hint from the state property when calling  `openFromState` and `createFromState`. This is because we may not have any login_hint cached but the login_hint will be provided in the state (via the userId property).

If we don't provide a login_hint in this situation then the user will unnecessarily be prompted with an auth dialog to pick which account--even though we already know which account the are accessing the document from.

begedin committed:
Moved fetchLoginHint in auth library to proper location (authorizeImmediate, not authorize)

begedin committed:
Removed duplication from login-hint, so a method is imported from uri library

begedin committed:
Added login hint to authenticate call from open* and createFromState

begedin committed:
Moved getDocumentIdFromLocation to uri library, since it's more fitting for it to be there

begedin commented:
@venkatd You're right. I completely missed that part. The simplest way to do it is to send in a login hint from the authenticate call in the `application-route-mixin`, where the document is first opened or created from state. We already have the state object there, so it's a simple matter if retrieving the `userID` from that and sending it as a `login_hint`.

That way, the session is still the thing that handles our authentication in all cases.

I also realized there was some code duplication in the login hint library, so I moved some stuff around to eliminate it. Basically, `login_hint` had an identical method to one being exported from the `uri` library. Additionally, another method `getDocumentIdFromLocation` is also something that fits with the `uri` library more than it does with `login_hint`, so I moved it and am now importing it from there.

venkatd commented on addon/lib/auth.js:
```diff
@@ -42,8 +45,9 @@ var Auth = Ember.Object.extend({
 
   authorizeImmediate: function(options) {
     return this.authorize(merge({
+      login_hint: fetchLoginHint(),
```
Is this the change you made?

venkatd commented on addon/mixins/application-route-mixin.js:
```diff
@@ -11,20 +11,17 @@ export default Ember.Mixin.create(ApplicationRouteMixin, {
-    if (!isStatePresent) {
-      return;
+    if (isStatePresent) {
+      this.get('session').authenticate('authenticator:gdrive', { 'login_hint': state.get('userID') }).then(function () {
```
Oh here also. Never mind I see it. I think previous versions of ember-simple-auth didn't support the 2nd param.
